File: Capitulo 1 - EDA/Ejemplo_estimaciones.py

# %%

import matplotlib.pylab as plt
import numpy as np
import os
import pandas as pd 
import wquantiles
import seaborn as sns

from statsmodels import robust
from scipy import stats

#%%

import common
from pathlib import Path

#DATA = common.dataDirectory()
DATA = Path().resolve() / 'data'
DATA

# %%

p = os.path.abspath(os.path.join(os.path.dirname(__file__),'..','data'))
print(p)

# %%

state  = pd.read_csv(p+'\state.csv')
state

# %%

# Media 
state['Población'].mean()

# %%

# Media truncada
stats.trim_mean(state['Población'], 0.1) # trim ignora el 10 de cada extremo

# %%

# Mediana
state['Población'].median()

# %%

# Media ponderada
np.average(state['Homicidios'], weights=state['Población'])

# %%

# Mediana ponderada
wquantiles.median(state['Homicidios'], weights=state['Población'])

# %%

# desvición estándar
state['Población'].std()
# %%

# IQR
state['Población'].quantile(0.75) - state['Población'].quantile(0.25)

# %%

# MAD
robust.scale.mad(state['Población'])

# %%

# Percentiles
state['Homicidios'].quantile([0.05, 0.25, 0.5, 0.75, 0.95])


# %%

# Tabla, Percentiles de tasa de omicidios por estados
percentages = [0.05, 0.25, 0.5, 0.75, 0.95]
df = pd.DataFrame(state['Homicidios'].quantile(percentages))
df.index = [f'{p * 100}%' for p in percentages]
print(df.transpose())

# %%

#Boxplot
ax = (state['Población']/1_000_000).plot.box(figsize=(3, 4))
ax.set_ylabel('Population (millions)')

plt.tight_layout()
plt.show()


# %%

# Tabla de frecuencias
binnedPopulation = pd.cut(state['Población'], 10)
binnedPopulation

# %%

binnedPopulation.value_counts()

# %%

binnedPopulation.name = 'binnedPopulation'
df = pd.concat([state, binnedPopulation], axis=1)
df = df.sort_values(by='Población')

groups = []
for group, subset in df.groupby(by='binnedPopulation'):
    groups.append({
        'BinRange': group,
        'Count': len(subset),
        'States': ','.join(subset.Abrebiatura)
    })

print(pd.DataFrame(groups))


# %%

# Histograma

ax = (state['Población']/1_000_000).plot.hist(figsize=(4,4))
ax.set_xlabel('Población (millions)')

# %%

ax = state['Homicidios'].plot.hist(density=True, xlim=[0, 12], bins=(range(1,12)))
state['Homicidios'].plot.density(ax=ax)
ax.set_xlabel('Murder rate (100.000)')

# %%

dfw  = pd.read_csv(p+'\dfw_airline.csv')
dfw

# %%

ax = dfw.transpose().plot.bar(figsize=(4, 4), legend=False)
ax.set_xlabel('Cause of delay')
ax.set_ylabel('Count')

# %%

sp500_sym = pd.read_csv(p+'\sp500_sectors.csv')
sp500_px = pd.read_csv(p+'\sp500_data.csv.gz', index_col=0)

# %%
telecomSymbols = sp500_sym[sp500_sym['sector'] == 'telecommunications_services']['symbol']

telecom = sp500_px.loc[sp500_px.index >= '2012-07-01', telecomSymbols]
telecom.corr()
print(telecom)

# %%

etfs = sp500_px.loc[sp500_px.index > '2012-07-01', 
                    sp500_sym[sp500_sym['sector'] == 'etf']['symbol']]
print(etfs.head())

# %%

fig, ax = plt.subplots(figsize=(5, 4))
ax = sns.heatmap(etfs.corr(), vmin=-1, vmax=1, 
                 cmap=sns.diverging_palette(20, 220, as_cmap=True),
                 ax=ax)

plt.tight_layout()
plt.show()


# %%

File: Capitulo 1 - EDA/Ejemplo_estimaciones.r
library(readr)

# root
print(getwd())

s = paste(getwd(),"/data/state.csv", sep="")
print(s)

p = "C:/Users/gurox/Estudios/Python/Estadística Práctica Python y R/data/state.csv"
print(p)

#install.packages("str_diff")
#library(str_diff)

s == p
tolower(p) == tolower(s)

# state  <- read.csv(s)
state <- read_csv(p)
state

mean(state[['Población']])

# Media truncada
mean(state[['Población']], trim=0.1) # trim ignora el 10 de cada extremo

median(state[['Población']])

#install.packages("matrixStats")
library("matrixStats")

# Media ponderada
weighted.mean(state[['Homicidios']], w=state[['Población']])

# Mediana ponderada
weightedMedian(state[['Homicidios']], w=state[['Población']])

# Desviación estandar
sd(state[['Población']])

# Rango Intercuartilico
IQR(state[['Población']])

#MAD
mad(state[['Población']])

# Percentiles
quantile(state[['Homicidios']], p=c(.05, .25, .5, .75, .95))

#Boxplot
boxplot(state[['Población']]/1000000, ylab='Population (millions)')

# Tabla de frecuencias
breaks <- seq(from=min(state[['Población']]), to=max(state[['Población']]), length=11)
breaks

pop_freq <- cut(state[['Población']], breaks=breaks, right=TRUE, include.lowest=TRUE)
pop_freq

table(pop_freq)

# Histogram
options(scipen=5)
hist(state[['Población']], breaks=breaks)

# Density
hist(state[['Población']], freq=FALSE)
lines(density(state[['Homicidios']]), lwd=3, col='blue')

# Data
s = paste(getwd(),"/data/dfw_airline.csv", sep="")
print(s)
dfw <- read.csv(s)
dfw

# Diagrama de barras
barplot(as.matrix(dfw)/6, ces.axis=0.8, cex.names=0.7, xlab='Cause of delay', ylab='Count')

# Correlación
PSDS_PATH <- file.path(dirname(dirname(getwd())))
print(PSDS_PATH)

p = paste(getwd(),"/data/sp500_data.csv.gz", sep="")
sp500_px <- read.csv(p, row.names=1)

p = paste(getwd(),"/data/sp500_sectors.csv", sep="")
sp500_sym <- read.csv(p, stringsAsFactors = FALSE)

# Correlación entre variables

telecom <- sp500_px[, sp500_sym[sp500_sym$sector == 'telecommunications_services', 'symbol']]
telecom <- telecom[row.names(telecom) > '2012-07-01',]
telecom_cor <- cor(telecom)
telecom_cor

# Diagrama de correlación

etfs <- sp500_px[row.names(sp500_px) > '2012-07-01', 
                 sp500_sym[sp500_sym$sector == 'etf', 'symbol']]

#install.packages("corrplot")
library('corrplot')

corrplot(cor(etfs), method='ellipse')

# Diagrama de dispersión

plot( telecom$T, telecom$VZ, xlab="ATT (T)", ylab="Verizon (VZ)")

